shortfin/python/shortfin_apps/sd/components/service.py

```diff
@@ -0,0 +1,330 @@
+                BatcherProcess.STROBE_SHORT_DELAY
+                if len(self.pending_preps) > 0
+                else BatcherProcess.STROBE_LONG_DELAY
+            )
+            if self.strobe_enabled:
+                self.submit(StrobeMessage())
+
+    async def run(self):
+        strober_task = asyncio.create_task(self._background_strober())
+        reader = self.batcher_infeed.reader()
+        while item := await reader():
+            self.strobe_enabled = False
+            if isinstance(item, InferenceExecRequest):
+                phase = item.phase
+                if phase == InferencePhase.PREP:
+                    self.pending_preps.add(item)
+                elif phase == InferencePhase.ENCODE:
+                    self.pending_encodes.add(item)
+                elif phase == InferencePhase.DENOISE:
+                    self.pending_denoises.add(item)
+                elif phase == InferencePhase.DECODE:
+                    self.pending_decodes.add(item)
+                elif phase == InferencePhase.POSTPROCESS:
+                    self.pending_postprocesses.add(item)
+                else:
+                    logger.error("Illegal InferenceExecRequest phase: %r", phase)
+            elif isinstance(item, StrobeMessage):
+                self.strobes += 1
+            else:
+                logger.error("Illegal message received by batcher: %r", item)
+            self.board_flights()
+            self.strobe_enabled = True
+        await strober_task
+
+    def board_flights(self):
+        waiting_count = (
+            len(self.pending_preps)
+            + len(self.pending_encodes)
+            + len(self.pending_denoises)
+            + len(self.pending_decodes)
+            + len(self.pending_postprocesses)
+        )
+        if waiting_count == 0:
+            return
+        if waiting_count < self.ideal_batch_size and self.strobes < 2:
+            logger.info("Waiting a bit longer to fill flight")
+            return
+        self.strobes = 0
+
+        self.board_preps()
+        self.board_encodes()
+        self.board_denoises()
+        self.board_decodes()
+        self.board_postprocesses()
+
+        # For now, kill anything that is left.
+        for i in [
+            self.pending_preps,
+            self.pending_encodes,
+            self.pending_denoises,
+            self.pending_decodes,
+            self.pending_postprocesses,
+        ]:
+            for request in i:
+                request.done.set_success()
+            i.clear()
+
+    def board_preps():
+        pass
+
+    def board_encodes():
+        pass
+
+    def board_denoises():
+        pass
+
+    def board_decodes():
+        pass
+
+    def board_postprocesses():
+        pass
+
+
+########################################################################################
+# Inference Executors
+########################################################################################
+
+
+class InferenceExecProcessPrep(sf.Process):
```

Comment:
If you took my advice above, there would just be one InferenceExec and it would do if req.need_*: inside.

Reply:
I think regardless of whether we do the finer-grained batching, it might not be best practice to split these processes up into components. I think it makes the implementation a bit tighter -- this single process structure accounts for any reasonable extension of the SD pipeline, barring expansion into significantly larger workflows like creative upscaling, but at that point we'd be out of implementation scope, I'd say. For now.